### PR TITLE
[DRAFT] pytests for losses

### PR DESCRIPTION
**Pull request by JINO-ROHIT**
Im writing this PR to help write test cases for all the losses .

Im starting with ContrastiveLoss and I wanted to get some early feedback on this. Like you mentioned, the idea is have a single parametrized func and pass all loss cases into it.

wdyt
@tomaarsen

### CODE REVIEW

**Comment by tomaarsen** [comment]
Hello!

I think this is heading in the right direction, but I do have some suggestions:

In particular, I think there's a risk that our test implementation has similar/the same bug as our real implementation. However, we still want to ensure that the loss "works".

So, one option is to create 2 batches. If we want to use the same model for both, then we have to create 1 "good" batch and 1 "bad" batch. The former is normal data, whereas the latter is opposite of reality: the positive is the negative, or related texts are marked as 0.0 similarity, etc.

Then, the same trained model should give a low loss for the "good" batch and a high loss for the "bad" batch. We should in both cases still test that the output 1) is a torch Tensor, 2) not `nan`, 3) not `0.0`, 4) has `requires_grad`, etc.

I think this is safer than just calculating an "expected loss", because if our implementation is buggy, that expected loss is probably also wrong.

And then we'd have to create the good and bad batches for each row in the [Loss Overview](https://sbert.net/docs/sentence_transformer/loss_overview.html), and then use each of the appropriate loss functions with those good/bad batches and ensure that `loss(good_batch) < loss(bad_batch)`.

Does that make sense?

- Tom Aarsen

**Comment by JINO-ROHIT** [comment]
umm i must admit i dont completely understand, do you mind giving an example?

**Comment by JINO-ROHIT** [comment]
yeap, thanks so much tom, working on it!

**Comment by tomaarsen** [comment]
I think there is a way for `pytest` to load a fixture by name, but I think it's almost as hacky as this, so I think this is fine to keep for now! If I ever run into it again, then I can update things.

**Comment by JINO-ROHIT** [comment]
cool then , ill add for other losses

**Comment by JINO-ROHIT** [comment]
wb for losses like softmax? 
this condition fails
```
assert correct_loss < incorrect_loss, "Loss should be lower for correct data than for incorrect data"
```
I assume its since the test tiny model hasnt been trained and the random init makes the losses inconsistent ?

**Comment by tomaarsen** [comment]
Oh, I hadn't thought about that the tiny model might not be trained enough, hah. 
I think I did train them a bit, perhaps you can check if this one does work? https://huggingface.co/sentence-transformers-testing/all-nli-bert-tiny-dense

And otherwise we'll have to use a bigger, normal model, like all-MiniLM-L6-v2.

- Tom Aarsen

**Comment by JINO-ROHIT** [comment]
yeah, this models seems to be working for now. thanks!

**Comment by JINO-ROHIT** [comment]
tried using a larger model and the smaller ones as well, the losses are still kinda random.

Also i think ive added the losses that can be tested together, i think the other losses have to be handled differently, wdyt?

**Comment by tomaarsen** [comment]
Hello!

This was tagged DRAFT in early 2025 and the loss directory layout has changed substantially with the v5.4 refactor (per-archetype subpackages, snake_case filenames). I think I'll close this now as it's also stale.

- Tom Aarsen
